Update R to 4.2 & Bioconductor to 3.15

jashapiro opened the pull request:
**Issue Addressed**
closes #183

**What is the purpose of these changes?**

The main goal here is to bring this repo more in sync with `sc-data-integration`, incorporating any updates that have come along with a new release of bioconductor.

**What changes did you make?**

- Update the environment to usr R 4.2. Since some packages are not available in conda, those are now installed by R in the setup script.
- update the `renv.lock` file with new package versions and requirements. There are a few changes in which packages are installed, but those may reflect changes in dependencies. 
- updated docs to reflect new requirements as stored in the `renv.lock` file for non-conda installation. 
 

**Were there any other solutions that you tried?**

Not as now. I was thinking a bit about whether we should maybe retain the old renv.lock file to allow some backward compatibility, but I hadn't quite explored how to do that.

The main reason to do that would be for development on the server, where we currently have R4.1. But since we are recommending the conda environments, _running_ the workflow there should still work.


**Any comments, concerns, or questions important for reviewers**

As previously, if testing on an Apple Silicon Mac, you will want to run 
```
CONDA_SUBDIR=osx-64 snakemake  -c1 --use-conda  --conda-create-envs-only build_renv
```
to build the new environment.

On an Intel Mac, the first run of the workflow should install the new enviroment automatically.

Building the new environment may be slow, but maybe will use a cache if you are lucky.


**Checklist**

Place an `x` in all boxes that you have completed.

- [x] I have run the most recent version of the code
- [x] I have added all necessary documentation (if applicable)

sjspielman commented on README.md:
```diff
@@ -145,7 +145,7 @@ You can install pandoc system-wide by following [pandoc's instructions](https://
 
 If you are on an Apple Silicon (M1/M2/Arm) Mac and are not using Snakemake and `conda` to handle dependencies, you will need to be sure that you have the Intel version of R, as Bioconductor packages do not currently support the Arm architecture.
-Clicking [this link](https://cran.r-project.org/bin/macosx/base/R-4.1.3.pkg) will download the Intel version of R, version 4.1.3, and you can install R following installation instructions.
+Clicking [this link](https://cran.r-project.org/bin/macosx/base/R-4.2.1.pkg) will download the Intel version of R, version 4.2.1, and you can install R following installation instructions.
```
I feel like this needs some more words

```suggestion
Clicking [this link](https://cran.r-project.org/bin/macosx/base/R-4.2.1.pkg) will download the Intel version of R, version 4.2.1, and you can install R by following the installation instructions.
```

sjspielman commented on utils/setup-functions.R:
```diff
@@ -35,14 +35,14 @@ check_r_bioc_versions <- function() {
-    stop("R version must be at least 4.1.")
+  # Check that R version is at least 4.2
+  if (! (R.version$major == 4 && R.version$minor >= 2)){
+    stop("R version must be at least 4.2.")
   }
 
   # Check that Bioconductor version is at least 3.14
-  if (packageVersion("BiocVersion") < 3.14){
-    stop("Bioconductor version is less than 3.14.")
+  if (packageVersion("BiocVersion") < 3.15){
+    stop("Bioconductor version is less than 3.15.")
```
Yes, I knew this function would come in handy 🎉 !

sjspielman commented on utils/setup-functions.R:
```diff
@@ -35,14 +35,14 @@ check_r_bioc_versions <- function() {
+    stop("R version must be at least 4.2.")
   }
 
   # Check that Bioconductor version is at least 3.14
```
```suggestion
  # Check that Bioconductor version is at least 3.15
```

sjspielman commented on envs/scpca-renv.yaml:
```diff
@@ -4,9 +4,9 @@ channels:
   - bioconda
   - defaults
 dependencies:
+  - cmake
```
Curious what requires `cmake` here, since we didn't need in the integration repo.

jashapiro replied:
I don't remember exactly, but one of the R packages did.

sjspielman reviewed (commented):
This line in the README needs to be updated too - 
https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/531f6f680f97a64ee5eb3bfdc6e58d88ea7053da/README.md?plain=1#L108

cbethell commented:
Interestingly, I checked out the branch locally and am getting the following error when I run `CONDA_SUBDIR=osx-64 snakemake  -c1 --use-conda  --conda-create-envs-only build_renv` for testing:
<img width="1148" alt="Screen Shot 2022-08-19 at 1 39 04 PM" src="https://user-images.githubusercontent.com/43576623/185685885-32c42239-7563-4fb7-9c93-068ee4bb1d26.png">

I tried scrapping things and starting from scratch but was not quite successful. Any idea what may be occurring here?

jashapiro commented:
> Interestingly, I checked out the branch locally and am getting the following error when I run `CONDA_SUBDIR=osx-64 snakemake -c1 --use-conda --conda-create-envs-only build_renv` for testing: <img alt="Screen Shot 2022-08-19 at 1 39 04 PM" width="1148" src="https://user-images.githubusercontent.com/43576623/185685885-32c42239-7563-4fb7-9c93-068ee4bb1d26.png">
> 
> I tried scrapping things and starting from scratch but was not quite successful. Any idea what may be occurring here?

@cbethell It looks like the installation was working fine, but then the next session couldn't find the packages, which is very strange! I updated the post-install script to do the installation of needed packages and bootstrapping `renv` all in a single session, which hopefully will fix the error you ran into.  Can you try running `CONDA_SUBDIR=osx-64 snakemake -c1 --use-conda --conda-create-envs-only build_renv` with the latest changes?

jashapiro commented:
@cbethell Looking at your error, there is something else very odd there, which is that it seems like the first installation tried to use the library at `renv/R-4.1/`, which it really should not be doing, since the version of R at that point should be 4.2 based on the data files there. Do you have anything set in an `.Renviron` file or other R config files in the `scpca-downstream-analysis` directory? I could also try switching from `Rscript --no-init-file` to `Rscript --vanilla` which would ignore whatever settings might be there, but if there are settings there that were interfering, they could cause downstream problems as well.

allyhawkins reviewed (dismissed):
Was able to test this on my intel mac and everything worked as expected and looks good 👍

cbethell reviewed (dismissed):
> @cbethell Looking at your error, there is something else very odd there, which is that it seems like the first installation tried to use the library at `renv/R-4.1/`, which it really should not be doing, since the version of R at that point should be 4.2 based on the data files there. Do you have anything set in an `.Renviron` file or other R config files in the `scpca-downstream-analysis` directory? I could also try switching from `Rscript --no-init-file` to `Rscript --vanilla` which would ignore whatever settings might be there, but if there are settings there that were interfering, they could cause downstream problems as well, so it would be worth tracking them down.

Looks like the latest changes interestingly worked for me (very strange behavior), but I was able to successfully test things locally! ✅

sjspielman commented:
I was getting the same error as @cbethell. I have all the recent changes pulled down, and I started with an OK `renv::status()`. I have no other config files. It's very unclear to me why it's not finding `jsonlite` and `purrr`. I've copied the relevant bits of the log, which suggests that this may be emerging from some `rlang` business...? 

```
installing to /Users/sjspielman/Library/R/x86_64/4.2/library/00LOCK-purrr/00new/purrr/libs\n** R\n** inst\n** byte-compile and prepare package for lazy loading\nError in library.dynam(lib, package, package.lib) : \n  shared object \xe2\x80\x98rlang.dylib\xe2\x80\x99 not found\nCalls: <Anonymous> ... loadNamespace -> namespaceImport -> loadNamespace -> library.dynam\nExecution halted\nERROR: lazy loading failed for package \xe2\x80\x98purrr\xe2\x80\x99\n* removing \xe2\x80\x98/Users/sjspielman/Library/R/x86_64/4.2/library/purrr\xe2\x80\x99\n* restoring previous \xe2\x80\x98/Users/sjspielman/Library/R/x86_64/4.2/library/purrr\xe2\x80\x99\n* installing *source* package \xe2\x80\x98renv\xe2\x80\x99 ...\n** package \xe2\x80\x98renv\xe2\x80\x99 successfully unpacked and MD5 sums checked\n** using staged installation\n** R\n** inst\n** byte-compile and prepare package for lazy loading\n** help\n*** installing help indices\n*** copying figures\n** building package indices\n** installing vignettes\n** testing if installed package can be loaded from temporary location\n** testing if installed package can be loaded from final location\n** testing if installed package keeps a record of temporary installation path\n* DONE (renv)\n\nThe downloaded source packages are in\n\t\xe2\x80\x98/private/var/folders/m7/gnk_vpg14hdbf06y8q6ppcyh0000gn/T/RtmpC2mP7L/downloaded_packages\xe2\x80\x99\nWarning message:\nIn install.packages(c("jsonlite", "purrr", "renv"), repos = "https://cloud.r-project.org") :\n  installation of package \xe2\x80\x98purrr\xe2\x80\x99 had non-zero exit status\nError: \nExecution halted\nError: \nExecution halted\n'
```

jashapiro commented:
> I was getting the same error as @cbethell. I have all the recent changes pulled down, and I started with an OK `renv::status()`. I have no other config files. It's very unclear to me why it's not finding `jsonlite` and `purrr`. I've copied the relevant bits of the log, which suggests that this may be emerging from some `rlang` business...?
> 
> ```
> installing to /Users/sjspielman/Library/R/x86_64/4.2/library/00LOCK-purrr/00new/purrr/libs\n** R\n** inst\n** byte-compile and prepare package for lazy loading\nError in library.dynam(lib, package, package.lib) : \n  shared object \xe2\x80\x98rlang.dylib\xe2\x80\x99 not found\nCalls: <Anonymous> ... loadNamespace -> namespaceImport -> loadNamespace -> library.dynam\nExecution halted\nERROR: lazy loading failed for package \xe2\x80\x98purrr\xe2\x80\x99\n* removing \xe2\x80\x98/Users/sjspielman/Library/R/x86_64/4.2/library/purrr\xe2\x80\x99\n* restoring previous \xe2\x80\x98/Users/sjspielman/Library/R/x86_64/4.2/library/purrr\xe2\x80\x99\n* installing *source* package \xe2\x80\x98renv\xe2\x80\x99 ...\n** package \xe2\x80\x98renv\xe2\x80\x99 successfully unpacked and MD5 sums checked\n** using staged installation\n** R\n** inst\n** byte-compile and prepare package for lazy loading\n** help\n*** installing help indices\n*** copying figures\n** building package indices\n** installing vignettes\n** testing if installed package can be loaded from temporary location\n** testing if installed package can be loaded from final location\n** testing if installed package keeps a record of temporary installation path\n* DONE (renv)\n\nThe downloaded source packages are in\n\t\xe2\x80\x98/private/var/folders/m7/gnk_vpg14hdbf06y8q6ppcyh0000gn/T/RtmpC2mP7L/downloaded_packages\xe2\x80\x99\nWarning message:\nIn install.packages(c("jsonlite", "purrr", "renv"), repos = "https://cloud.r-project.org") :\n  installation of package \xe2\x80\x98purrr\xe2\x80\x99 had non-zero exit status\nError: \nExecution halted\nError: \nExecution halted\n'
> ```

Are you getting this error now, or does it work with the changes I made?

sjspielman commented:
> Are you getting this error now, or does it work with the changes I made?

Now, with my `git log`  showing the most recent commit as 4670c55

jashapiro commented:
The fact that it is apparently installing to `/00LOCK-purrr/00new` is very strange. I would also not expect it to be installing to your user library at all. 

Can you try it with substituting `Rscript --vanilla` in the `post-deploy` script?

sjspielman commented:
> Can you try it with substituting `Rscript --vanilla` in the post-deploy script?

I probably should have said that I did that already, and it didn't work. I also tried to separate out the different `Rscript` lines, and I got past the installation of `jsonlite` and `purrr` but then it failed.

jashapiro commented:
@sjspielman I just pushed a version of the installation script that will write out the Rscript output to `rinstall_log.txt`. 
I also included `setup.sh` that calls the expected snakemake commands on Apple Silcon.
Can you run that and send me the output?

sjspielman commented:
Enjoy! Looks indeed `rlang`-y, maybe..
[rinstall_log.txt](https://github.com/AlexsLemonade/scpca-downstream-analyses/files/9396950/rinstall_log.txt)

jashapiro commented:
> Enjoy! Looks indeed `rlang`-y, maybe.. [rinstall_log.txt](https://github.com/AlexsLemonade/scpca-downstream-analyses/files/9396950/rinstall_log.txt)

@sjspielman I think it is actually the installation location; it should not be installing in your global user lib, so let's try unsetting that. Changes pushed.

sjspielman commented:
Same situation, and with this at the top - 
```
Installing packages into ‘/Users/sjspielman/Library/R/x86_64/4.2/library’
(as ‘lib’ is unspecified)
```

Is there a world where snakemake gives us access to the hash of the directory where we _would_ like to install into, and we can pop that in as a `lib` argument? Like, `.snakemake/conda/$hash`?
[rinstall_log.txt](https://github.com/AlexsLemonade/scpca-downstream-analyses/files/9397193/rinstall_log.txt)

sjspielman commented:
I ran this also without the post deploy just to make sure R was properly installed in the right location. I confirmed it's definitely using the environment-installed R and not regular R build - 
```
(snakemake) sjs-ccdl :: ALSF/scpca/scpca-downstream-analyses ‹jashapiro/r4.2*› » conda activate /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4
(/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4) sjs-ccdl :: ALSF/scpca/scpca-downstream-analyses ‹jashapiro/r4.2*› » which R
/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/R
```

jashapiro commented:
> Same situation, and with this at the top -
> 
> ```
> Installing packages into ‘/Users/sjspielman/Library/R/x86_64/4.2/library’
> (as ‘lib’ is unspecified)
> ```
> 
> Is there a world where snakemake gives us access to the hash of the directory where we _would_ like to install into, and we can pop that in as a `lib` argument? Like, `.snakemake/conda/$hash`?

It should be set automatically, which is what is so confusing. I just added printing `.libPaths()` as the first step of the script. And after unsetting `R_LIBS_USER`, it should not have found the library it did in `/Users/sjspielman/Library/R/`, so that is what I am focused on for the moment.

jashapiro commented:
> I ran this also without the post deploy just to make sure R was properly installed in the right location. I confirmed it's definitely using the environment-installed R and not regular R build -

From this `R` can you try running `.libPaths()` and and `Sys.getenv()`?

sjspielman commented:
Just checked that it's definitely set automatically! I ran with an added `conda info --envs` at the top of the post-deploy, and the log indeed shows the right environment is on..

sjspielman commented:
I think it might be this one `RENV_DEFAULT_R_LIBS_USER`. I'll try unsetting it.

```
> .libPaths()
[1] "/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/renv/library/R-4.2/x86_64-apple-darwin13.4.0"                   
[2] "/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/library"
> Sys.getenv()
__CF_USER_TEXT_ENCODING
                        0x1F5:0x0:0x0
__CFBundleIdentifier    com.apple.Terminal
_CE_CONDA               
_CE_M                   
_CONDA_PYTHON_SYSCONFIGDATA_NAME
                        _sysconfigdata_x86_64_apple_darwin13_4_0
ac_cv_func_malloc_0_nonnull
                        yes
ac_cv_func_realloc_0_nonnull
                        yes
AR                      x86_64-apple-darwin13.4.0-ar
AS                      x86_64-apple-darwin13.4.0-as
BUILD                   x86_64-apple-darwin13.4.0
build_alias             x86_64-apple-darwin13.4.0
CC                      x86_64-apple-darwin13.4.0-clang
CC_FOR_BUILD            /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-clang
CFLAGS                  -march=core2 -mtune=haswell -mssse3
                        -ftree-vectorize -fPIC -fPIE
                        -fstack-protector-strong -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
CHECKSYMS               x86_64-apple-darwin13.4.0-checksyms
CLANG                   x86_64-apple-darwin13.4.0-clang
CLANGXX                 x86_64-apple-darwin13.4.0-clang++
CMAKE_ARGS              -DCMAKE_AR=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ar
                        -DCMAKE_CXX_COMPILER_AR=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ar
                        -DCMAKE_C_COMPILER_AR=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ar
                        -DCMAKE_RANLIB=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ranlib
                        -DCMAKE_CXX_COMPILER_RANLIB=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ranlib
                        -DCMAKE_C_COMPILER_RANLIB=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ranlib
                        -DCMAKE_LINKER=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ld
                        -DCMAKE_STRIP=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-strip
                        -DCMAKE_INSTALL_NAME_TOOL=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-install_name_tool
                        -DCMAKE_LIBTOOL=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-libtool
                        -DCMAKE_OSX_DEPLOYMENT_TARGET=
                        -DCMAKE_BUILD_TYPE=Release
                        -DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
CMAKE_PREFIX_PATH       :/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4
COLUMNS                 141
CONDA_BACKUP_HOST       sjs-ccdl.local
CONDA_DEFAULT_ENV       /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4
CONDA_EXE               /Users/sjspielman/opt/miniconda3/bin/conda
CONDA_PREFIX            /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4
CONDA_PREFIX_1          /Users/sjspielman/opt/miniconda3
CONDA_PREFIX_2          /Users/sjspielman/opt/miniconda3/envs/snakemake
CONDA_PROMPT_MODIFIER   (/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4)
CONDA_PYTHON_EXE        /Users/sjspielman/opt/miniconda3/bin/python
CONDA_SHLVL             3
CONDA_TOOLCHAIN_BUILD   x86_64-apple-darwin13.4.0
CONDA_TOOLCHAIN_HOST    x86_64-apple-darwin13.4.0
CPPFLAGS                -D_FORTIFY_SOURCE=2 -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
CXX                     x86_64-apple-darwin13.4.0-clang++
CXX_FOR_BUILD           /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-clang++
CXXFLAGS                -march=core2 -mtune=haswell -mssse3
                        -ftree-vectorize -fPIC -fPIE
                        -fstack-protector-strong -O2 -pipe
                        -stdlib=libc++ -fvisibility-inlines-hidden
                        -fmessage-length=0 -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
DEBUG_CFLAGS            -march=core2 -mtune=haswell -mssse3
                        -ftree-vectorize -fPIC -fPIE
                        -fstack-protector-strong -O2 -pipe -Og -g -Wall
                        -Wextra -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
DEBUG_CXXFLAGS          -march=core2 -mtune=haswell -mssse3
                        -ftree-vectorize -fPIC -fPIE
                        -fstack-protector-strong -O2 -pipe
                        -stdlib=libc++ -fvisibility-inlines-hidden
                        -fmessage-length=0 -Og -g -Wall -Wextra
                        -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
DEBUG_FFLAGS            -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
                        -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -Og -g -Wall
                        -Wextra -fcheck=all -fbacktrace -fimplicit-none
                        -fvar-tracking-assignments
DEBUG_FORTRANFLAGS      -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
                        -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -Og -g -Wall
                        -Wextra -fcheck=all -fbacktrace -fimplicit-none
                        -fvar-tracking-assignments
DISPLAY                 /private/tmp/com.apple.launchd.4ZxJtzZVcj/org.xquartz:0
DYLD_FALLBACK_LIBRARY_PATH
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/lib:/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib:/Users/runner/miniforge3/conda-bld/r-base-split_1659993936932/_build_env/lib/server
EDITOR                  vi
F77                     /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
F90                     /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
F95                     /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
FC                      /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
FC_FOR_BUILD            /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
FFLAGS                  -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
FORTRANFLAGS            -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
GEM_HOME                /Users/sjspielman/.gem/ruby/3.1.2
GEM_PATH                /Users/sjspielman/.gem/ruby/3.1.2:/Users/sjspielman/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0
GEM_ROOT                /Users/sjspielman/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0
GFORTRAN                /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
GSETTINGS_SCHEMA_DIR    /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/share/glib-2.0/schemas
GSETTINGS_SCHEMA_DIR_CONDA_BACKUP
                        
HOME                    /Users/sjspielman
HOMEBREW_CELLAR         /opt/homebrew/Cellar
HOMEBREW_PREFIX         /opt/homebrew
HOMEBREW_REPOSITORY     /opt/homebrew
HOST                    x86_64-apple-darwin13.4.0
host_alias              x86_64-apple-darwin13.4.0
INFOPATH                /opt/homebrew/share/info:
INSTALL_NAME_TOOL       x86_64-apple-darwin13.4.0-install_name_tool
LANG                    en_US.UTF-8
LD                      x86_64-apple-darwin13.4.0-ld
LDFLAGS                 -Wl,-pie -Wl,-headerpad_max_install_names
                        -Wl,-dead_strip_dylibs
                        -Wl,-rpath,/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib
                        -L/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib
LDFLAGS_LD              -pie -headerpad_max_install_names
                        -dead_strip_dylibs -rpath
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib
                        -L/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib
LESS                    -R
LIBTOOL                 x86_64-apple-darwin13.4.0-libtool
LINES                   21
LIPO                    x86_64-apple-darwin13.4.0-lipo
LN_S                    ln -s
LOGNAME                 sjspielman
LSCOLORS                Gxfxcxdxbxegedabagacad
MAGICK_HOME             /Users/sjspielman/bin/ImageMagick-7.0.10
MAKE                    make
MANPATH                 /opt/homebrew/share/man::
MESON_ARGS              --buildtype release
NM                      x86_64-apple-darwin13.4.0-nm
NMEDIT                  x86_64-apple-darwin13.4.0-nmedit
OBJC                    x86_64-apple-darwin13.4.0-clang
OBJC_FOR_BUILD          /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-clang
OTOOL                   x86_64-apple-darwin13.4.0-otool
PAGER                   less
PAGESTUFF               x86_64-apple-darwin13.4.0-pagestuff
PATH                    /Users/sjspielman/.gem/ruby/3.1.2/bin:/Users/sjspielman/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/bin:/Users/sjspielman/.rubies/ruby-3.1.2/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin:/Users/sjspielman/opt/miniconda3/condabin:/Users/sjspielman/.gem/ruby/2.7.0/bin:/usr/local/opt/ruby/bin:/Users/sjspielman/bin/ImageMagick-7.0.10/bin:/Users/sjspielman/bin:/usr/local/bin:/usr/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/opt/X11/bin:/opt/R/arm64/gfortran/bin
PWD                     /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses
R_ARCH                  
R_BROWSER               /usr/bin/open
R_BZIPCMD               /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/bzip2
R_DOC_DIR               /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/doc
R_GZIPCMD               /usr/bin/gzip
R_HOME                  /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R
R_INCLUDE_DIR           /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/include
R_LIBS_SITE             /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/site-library
R_LIBS_USER             /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/renv/library/R-4.2/x86_64-apple-darwin13.4.0:/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/library
R_PAPERSIZE             a4
R_PDFVIEWER             /usr/bin/open
R_PLATFORM              x86_64-apple-darwin13.4.0
R_PRINTCMD              lpr
R_RD4PDF                times,inconsolata,hyper
R_SESSION_TMPDIR        /var/folders/m7/gnk_vpg14hdbf06y8q6ppcyh0000gn/T//RtmpzmQBWn
R_SHARE_DIR             /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/share
R_STRIP_SHARED_LIB      x86_64-apple-darwin13.4.0-strip -x
R_STRIP_STATIC_LIB      x86_64-apple-darwin13.4.0-strip -S
R_SYSTEM_ABI            macos,gcc,gxx,gfortran,gfortran
R_TEXI2DVICMD           /Users/runner/miniforge3/conda-bld/r-base-split_1659993936932/_build_env/bin/texi2dvi
R_UNZIPCMD              /usr/bin/unzip
R_ZIPCMD                /usr/bin/zip
RANLIB                  x86_64-apple-darwin13.4.0-ranlib
REDO_PREBINDING         x86_64-apple-darwin13.4.0-redo_prebinding
RENV_DEFAULT_R_ENVIRON
                        <NA>
RENV_DEFAULT_R_ENVIRON_USER
                        <NA>
RENV_DEFAULT_R_LIBS     <NA>
RENV_DEFAULT_R_LIBS_SITE
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/site-library
RENV_DEFAULT_R_LIBS_USER
                        /Users/sjspielman/Library/R/x86_64/4.2/library
RENV_DEFAULT_R_PROFILE
                        <NA>
RENV_DEFAULT_R_PROFILE_USER
                        <NA>
RENV_PROJECT            /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses
RETICULATE_MINICONDA_PYTHON_ENVPATH
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/renv/python/r-reticulate
RSTUDIO_PANDOC          /Users/sjspielman/opt/miniconda3/bin/
RSTUDIO_WHICH_R         /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/R
RUBY_ENGINE             ruby
RUBY_ROOT               /Users/sjspielman/.rubies/ruby-3.1.2
RUBY_VERSION            3.1.2
RUBYOPT                 
SDKROOT                 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
SED                     /usr/bin/sed
SEG_ADDR_TABLE          x86_64-apple-darwin13.4.0-seg_addr_table
SEG_HACK                x86_64-apple-darwin13.4.0-seg_hack
SEGEDIT                 x86_64-apple-darwin13.4.0-segedit
SHELL                   /bin/zsh
SHLVL                   1
SIZE                    x86_64-apple-darwin13.4.0-size
SSH_AUTH_SOCK           /private/tmp/com.apple.launchd.y8bynbxuiy/Listeners
STRINGS                 x86_64-apple-darwin13.4.0-strings
STRIP                   x86_64-apple-darwin13.4.0-strip
TAR                     /usr/bin/tar
TERM                    xterm-256color
TERM_PROGRAM            Apple_Terminal
TERM_PROGRAM_VERSION    445
TERM_SESSION_ID         E18AF3E4-9FA0-4F62-96C1-793C88FD5AA7
TMPDIR                  /var/folders/m7/gnk_vpg14hdbf06y8q6ppcyh0000gn/T/
USER                    sjspielman
XPC_FLAGS               0x0
XPC_SERVICE_NAME        0
ZSH                     /Users/sjspielman/.oh-my-zsh
```

jashapiro commented:
Sorry, I should have said to run that from `R --vanilla`. The renv stuff will not be set if the `.Rprofile` file is not sourced

jashapiro commented:
> PATH                    /Users/sjspielman/.gem/ruby/3.1.2/bin:/Users/sjspielman/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/bin:/Users/sjspielman/.rubies/ruby-3.1.2/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin:/Users/sjspielman/opt/miniconda3/condabin:/Users/sjspielman/.gem/ruby/2.7.0/bin:/usr/local/opt/ruby/bin:/Users/sjspielman/bin/ImageMagick-7.0.10/bin:/Users/sjspielman/bin:/usr/local/bin:/usr/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/opt/X11/bin:/opt/R/arm64/gfortran/bin

You have homebrew in your `PATH` _before_ conda, and I suspect that this is a source of some conflicts. conda should always be before homebrew, or you will not get the full benefit of switching environments.

sjspielman commented:
Yup, ok, `"/Users/sjspielman/Library/R/x86_64/4.2/library"` is the first entry in `.libPaths()` instead of the conda library path which is 2nd. It's also the value for `R_LIBS_USER`. 

```
> .libPaths()
[1] "/Users/sjspielman/Library/R/x86_64/4.2/library"                                                                        
[2] "/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/library"

> Sys.getenv()
__CF_USER_TEXT_ENCODING
                        0x1F5:0x0:0x0
__CFBundleIdentifier    com.apple.Terminal
_CE_CONDA               
_CE_M                   
_CONDA_PYTHON_SYSCONFIGDATA_NAME
                        _sysconfigdata_x86_64_apple_darwin13_4_0
ac_cv_func_malloc_0_nonnull
                        yes
ac_cv_func_realloc_0_nonnull
                        yes
AR                      x86_64-apple-darwin13.4.0-ar
AS                      x86_64-apple-darwin13.4.0-as
BUILD                   x86_64-apple-darwin13.4.0
build_alias             x86_64-apple-darwin13.4.0
CC                      x86_64-apple-darwin13.4.0-clang
CC_FOR_BUILD            /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-clang
CFLAGS                  -march=core2 -mtune=haswell -mssse3
                        -ftree-vectorize -fPIC -fPIE
                        -fstack-protector-strong -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
CHECKSYMS               x86_64-apple-darwin13.4.0-checksyms
CLANG                   x86_64-apple-darwin13.4.0-clang
CLANGXX                 x86_64-apple-darwin13.4.0-clang++
CMAKE_ARGS              -DCMAKE_AR=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ar
                        -DCMAKE_CXX_COMPILER_AR=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ar
                        -DCMAKE_C_COMPILER_AR=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ar
                        -DCMAKE_RANLIB=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ranlib
                        -DCMAKE_CXX_COMPILER_RANLIB=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ranlib
                        -DCMAKE_C_COMPILER_RANLIB=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ranlib
                        -DCMAKE_LINKER=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-ld
                        -DCMAKE_STRIP=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-strip
                        -DCMAKE_INSTALL_NAME_TOOL=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-install_name_tool
                        -DCMAKE_LIBTOOL=/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-libtool
                        -DCMAKE_OSX_DEPLOYMENT_TARGET=
                        -DCMAKE_BUILD_TYPE=Release
                        -DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
CMAKE_PREFIX_PATH       :/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4
COLUMNS                 155
CONDA_BACKUP_HOST       sjs-ccdl.local
CONDA_DEFAULT_ENV       /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4
CONDA_EXE               /Users/sjspielman/opt/miniconda3/bin/conda
CONDA_PREFIX            /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4
CONDA_PREFIX_1          /Users/sjspielman/opt/miniconda3
CONDA_PREFIX_2          /Users/sjspielman/opt/miniconda3/envs/snakemake
CONDA_PREFIX_3          /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4
CONDA_PREFIX_4          /Users/sjspielman/opt/miniconda3
CONDA_PREFIX_5          /Users/sjspielman/opt/miniconda3/envs/snakemake
CONDA_PROMPT_MODIFIER   (/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4)
CONDA_PYTHON_EXE        /Users/sjspielman/opt/miniconda3/bin/python
CONDA_SHLVL             6
CONDA_TOOLCHAIN_BUILD   x86_64-apple-darwin13.4.0
CONDA_TOOLCHAIN_HOST    x86_64-apple-darwin13.4.0
CPPFLAGS                -D_FORTIFY_SOURCE=2 -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
CXX                     x86_64-apple-darwin13.4.0-clang++
CXX_FOR_BUILD           /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-clang++
CXXFLAGS                -march=core2 -mtune=haswell -mssse3
                        -ftree-vectorize -fPIC -fPIE
                        -fstack-protector-strong -O2 -pipe
                        -stdlib=libc++ -fvisibility-inlines-hidden
                        -fmessage-length=0 -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
DEBUG_CFLAGS            -march=core2 -mtune=haswell -mssse3
                        -ftree-vectorize -fPIC -fPIE
                        -fstack-protector-strong -O2 -pipe -Og -g -Wall
                        -Wextra -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
DEBUG_CXXFLAGS          -march=core2 -mtune=haswell -mssse3
                        -ftree-vectorize -fPIC -fPIE
                        -fstack-protector-strong -O2 -pipe
                        -stdlib=libc++ -fvisibility-inlines-hidden
                        -fmessage-length=0 -Og -g -Wall -Wextra
                        -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
DEBUG_FFLAGS            -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
                        -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -Og -g -Wall
                        -Wextra -fcheck=all -fbacktrace -fimplicit-none
                        -fvar-tracking-assignments
DEBUG_FORTRANFLAGS      -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
                        -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -Og -g -Wall
                        -Wextra -fcheck=all -fbacktrace -fimplicit-none
                        -fvar-tracking-assignments
DISPLAY                 /private/tmp/com.apple.launchd.4ZxJtzZVcj/org.xquartz:0
DYLD_FALLBACK_LIBRARY_PATH
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/lib:/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib:/Users/runner/miniforge3/conda-bld/r-base-split_1659993936932/_build_env/lib/server
EDITOR                  vi
F77                     /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
F90                     /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
F95                     /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
FC                      /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
FC_FOR_BUILD            /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
FFLAGS                  -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
FORTRANFLAGS            -march=core2 -mtune=haswell -ftree-vectorize
                        -fPIC -fstack-protector -O2 -pipe -isystem
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/include
GEM_HOME                /Users/sjspielman/.gem/ruby/3.1.2
GEM_PATH                /Users/sjspielman/.gem/ruby/3.1.2:/Users/sjspielman/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0
GEM_ROOT                /Users/sjspielman/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0
GFORTRAN                /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-gfortran
GSETTINGS_SCHEMA_DIR    /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/share/glib-2.0/schemas
GSETTINGS_SCHEMA_DIR_CONDA_BACKUP
                        
HOME                    /Users/sjspielman
HOMEBREW_CELLAR         /opt/homebrew/Cellar
HOMEBREW_PREFIX         /opt/homebrew
HOMEBREW_REPOSITORY     /opt/homebrew
HOST                    x86_64-apple-darwin13.4.0
host_alias              x86_64-apple-darwin13.4.0
INFOPATH                /opt/homebrew/share/info:
INSTALL_NAME_TOOL       x86_64-apple-darwin13.4.0-install_name_tool
LANG                    en_US.UTF-8
LD                      x86_64-apple-darwin13.4.0-ld
LDFLAGS                 -Wl,-pie -Wl,-headerpad_max_install_names
                        -Wl,-dead_strip_dylibs
                        -Wl,-rpath,/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib
                        -L/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib
LDFLAGS_LD              -pie -headerpad_max_install_names
                        -dead_strip_dylibs -rpath
                        /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib
                        -L/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib
LESS                    -R
LIBTOOL                 x86_64-apple-darwin13.4.0-libtool
LINES                   29
LIPO                    x86_64-apple-darwin13.4.0-lipo
LN_S                    ln -s
LOGNAME                 sjspielman
LSCOLORS                Gxfxcxdxbxegedabagacad
MAGICK_HOME             /Users/sjspielman/bin/ImageMagick-7.0.10
MAKE                    make
MANPATH                 /opt/homebrew/share/man::
MESON_ARGS              --buildtype release
NM                      x86_64-apple-darwin13.4.0-nm
NMEDIT                  x86_64-apple-darwin13.4.0-nmedit
OBJC                    x86_64-apple-darwin13.4.0-clang
OBJC_FOR_BUILD          /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/x86_64-apple-darwin13.4.0-clang
OTOOL                   x86_64-apple-darwin13.4.0-otool
PAGER                   less
PAGESTUFF               x86_64-apple-darwin13.4.0-pagestuff
PATH                    /Users/sjspielman/.gem/ruby/3.1.2/bin:/Users/sjspielman/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/bin:/Users/sjspielman/.rubies/ruby-3.1.2/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin:/Users/sjspielman/opt/miniconda3/condabin:/Users/sjspielman/.gem/ruby/2.7.0/bin:/usr/local/opt/ruby/bin:/Users/sjspielman/bin/ImageMagick-7.0.10/bin:/Users/sjspielman/bin:/usr/local/bin:/usr/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/opt/X11/bin:/opt/R/arm64/gfortran/bin
PWD                     /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses
R_ARCH                  
R_BROWSER               /usr/bin/open
R_BZIPCMD               /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/bzip2
R_DOC_DIR               /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/doc
R_ENVIRON               
R_ENVIRON_USER          
R_GZIPCMD               /usr/bin/gzip
R_HOME                  /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R
R_INCLUDE_DIR           /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/include
R_LIBS_SITE             /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/site-library
R_LIBS_USER             /Users/sjspielman/Library/R/x86_64/4.2/library
R_PAPERSIZE             a4
R_PDFVIEWER             /usr/bin/open
R_PLATFORM              x86_64-apple-darwin13.4.0
R_PRINTCMD              lpr
R_PROFILE               
R_PROFILE_USER          
R_RD4PDF                times,inconsolata,hyper
R_SESSION_TMPDIR        /var/folders/m7/gnk_vpg14hdbf06y8q6ppcyh0000gn/T//RtmpPwBgkz
R_SHARE_DIR             /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/lib/R/share
R_STRIP_SHARED_LIB      x86_64-apple-darwin13.4.0-strip -x
R_STRIP_STATIC_LIB      x86_64-apple-darwin13.4.0-strip -S
R_SYSTEM_ABI            macos,gcc,gxx,gfortran,gfortran
R_TEXI2DVICMD           /Users/runner/miniforge3/conda-bld/r-base-split_1659993936932/_build_env/bin/texi2dvi
R_UNZIPCMD              /usr/bin/unzip
R_ZIPCMD                /usr/bin/zip
RANLIB                  x86_64-apple-darwin13.4.0-ranlib
REDO_PREBINDING         x86_64-apple-darwin13.4.0-redo_prebinding
RSTUDIO_PANDOC          /Users/sjspielman/opt/miniconda3/bin/
RSTUDIO_WHICH_R         /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/.snakemake/conda/4e16cda23ca2a59556714669fe76f3d4/bin/R
RUBY_ENGINE             ruby
RUBY_ROOT               /Users/sjspielman/.rubies/ruby-3.1.2
RUBY_VERSION            3.1.2
RUBYOPT                 
SDKROOT                 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
SED                     /usr/bin/sed
SEG_ADDR_TABLE          x86_64-apple-darwin13.4.0-seg_addr_table
SEG_HACK                x86_64-apple-darwin13.4.0-seg_hack
SEGEDIT                 x86_64-apple-darwin13.4.0-segedit
SHELL                   /bin/zsh
SHLVL                   1
SIZE                    x86_64-apple-darwin13.4.0-size
SSH_AUTH_SOCK           /private/tmp/com.apple.launchd.y8bynbxuiy/Listeners
STRINGS                 x86_64-apple-darwin13.4.0-strings
STRIP                   x86_64-apple-darwin13.4.0-strip
TAR                     /usr/bin/tar
TERM                    xterm-256color
TERM_PROGRAM            Apple_Terminal
TERM_PROGRAM_VERSION    445
TERM_SESSION_ID         A89EC04F-1A53-4552-B4B7-BCC203D6D92F
TMPDIR                  /var/folders/m7/gnk_vpg14hdbf06y8q6ppcyh0000gn/T/
USER                    sjspielman
XPC_FLAGS               0x0
XPC_SERVICE_NAME        0
ZSH                     /Users/sjspielman/.oh-my-zsh
```


I got closer to a successful build by doing this hack!
```
# install packages required for renv
Rscript --vanilla -e \
  "
   target_lib_path <- grep('snakemake', .libPaths(), value=TRUE)
   install.packages(c('jsonlite', 'purrr'), repos = 'https://cloud.r-project.org', lib = target_lib_path)
   source('.Rprofile')
   renv::restore()
  " &> rinstall_log.txt
```
[rinstall_log.txt](https://github.com/AlexsLemonade/scpca-downstream-analyses/files/9397450/rinstall_log.txt)
 
My current thought is `library.dynam()` is using the wrong library location, but I have 0 idea how to address that under-the-hood installation step.

sjspielman commented:
Solved!! Aka, I reached the step where the main `renv` stuff is all building now, as I have no cache. But I feel confident this will wrap up fine.  

The post-deploy needs to install more dependencies -  `rlang` and `magrittr`! The following worked for me. I haven't dug into this to see how it can be simplified, e.g. whether `rlang` or `magrittr` have to be in their own lines, and whether the target path variable is definitely needed. 
```
# install packages required for renv
Rscript --vanilla -e \
  "
   target_lib_path <- grep('snakemake', .libPaths(), value=TRUE)
   install.packages('magrittr',  repos = 'https://cloud.r-project.org', lib = target_lib_path)
    install.packages('rlang',  repos = 'https://cloud.r-project.org', lib = target_lib_path)
   install.packages(c('jsonlite', 'purrr'), repos = 'https://cloud.r-project.org', lib = target_lib_path)
   source('.Rprofile')
   renv::restore()
  " &> rinstall_log.txt
```

sjspielman commented:
Update: The original code without my hacking paths in there works _if and only if_ this directory does not exist on my computer: `~/Library/R/x86_64/4.2/`, which contained user-level libraries installed for my user version of 4.2, which is separate from that installed into the `conda` environment.

jashapiro commented:
> Update: The original code without my hacking paths in there works _if and only if_ this directory does not exist on my computer: `~/Library/R/x86_64/4.2/`, which contained user-level libraries installed for my user version of 4.2, which is separate from that installed into the `conda` environment.

Okay, that is helpful. I don't like the hacking, but I can work with this to find a more robust solution.

jashapiro commented:
> Update: The original code without my hacking paths in there works _if and only if_ this directory does not exist on my computer: `~/Library/R/x86_64/4.2/`, which contained user-level libraries installed for my user version of 4.2, which is separate from that installed into the `conda` environment.

Just to check, did you try this with the latest update that does include `magrittr` and `rlang` installations?

jashapiro commented:
I have now updated the installation script to make it a bit more robust and added its usage to the readme. This should simplify the user experience, as the script itself is testing the architecture, and users should be able to use the same command for setup no matter what their architecture.

I am dismissing reviews, and would appreciate another round of testing. Hopefully this should be as simple as running `bash setup_env.sh`.

sjspielman commented:
Hmm..
```
(snakemake) sjs-ccdl :: ALSF/scpca/scpca-downstream-analyses ‹jashapiro/r4.2› » bash setup_envs.sh      
On Apple Silicon: Building R environment and required R packages with CONDA_SUBDIR=osx-64 for compatibility
Building DAG of jobs...
MissingInputException in rule normalize_data in line 76 of /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/Snakefile:
Missing input files for rule normalize_data:
    output: example-results/sample01/library01_miQC_normalized.rds
    wildcards: basename=example-results/sample01/library01_miQC
    affected files:
        example-results/sample01/library01_miQC_filtered.rds
```

But if I remove the force option `-R` from the bash script on line 14 I have a successful build - 

```
 CONDA_SUBDIR=osx-64 snakemake --use-conda --conda-create-envs-only -c1 build_renv 
Building DAG of jobs...
Creating conda environment envs/scpca-renv.yaml...
Downloading and installing remote packages.
Running post-deploy script .snakemake/conda/880e6788549a79c8b24be7d60c94541b.post-deploy.sh...
Environment for /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/envs/scpca-renv.yaml created (location: .snakemake/conda/880e6788549a79c8b24be7d60c94541b)
```

I wonder if I have all the necessary local non-version-control files here. I do have a populated `example-data/` directory.

cbethell commented:
> Hmm..
> 
> ```
> (snakemake) sjs-ccdl :: ALSF/scpca/scpca-downstream-analyses ‹jashapiro/r4.2› » bash setup_envs.sh      
> On Apple Silicon: Building R environment and required R packages with CONDA_SUBDIR=osx-64 for compatibility
> Building DAG of jobs...
> MissingInputException in rule normalize_data in line 76 of /Users/sjspielman/ALSF/scpca/scpca-downstream-analyses/Snakefile:
> Missing input files for rule normalize_data:
>     output: example-results/sample01/library01_miQC_normalized.rds
>     wildcards: basename=example-results/sample01/library01_miQC
>     affected files:
>         example-results/sample01/library01_miQC_filtered.rds
> ```
> 

I was actually able to reproduce this error by removed the `example-results` files, fwiw (it seemed to work fine when `example-results` was already populated from a previous run).

allyhawkins commented:
> I was actually able to reproduce this error by removed the example-results files, fwiw (it seemed to work fine when example-results was already populated from a previous run).

Yea... I was wondering if this is because the `example_results` folder doesn't yet exist and the directory has to already exist before hand for snakemake to use it?

jashapiro commented:
> > I was actually able to reproduce this error by removed the example-results files, fwiw (it seemed to work fine when example-results was already populated from a previous run).
> 
> Yea... I was wondering if this is because the `example_results` folder doesn't yet exist and the directory has to already exist before hand for snakemake to use it?

Snakemake shouldn't need the directory; it will create it if needed based on the rules it is following. So the question is what rule does it think it is following that it needs that file and can't find what will create it...

jashapiro commented:
Okay, the path error should now be fixed. It was an extra slash, and snakemake apparently thinks that `dir//file` is different from `dir/file` (and os.path.dir doesn't removed extra slashes like I would have expected). I also removed the `-R`, which was not really having the effect that I was expecting (I would not have expected it to cause Snakemake to look at _other_ rules, which I really didn't want it to do).

cbethell reviewed (approved):
✅
